src/shared/protocol.ts

```diff
@@ -995,39 +995,35 @@ export abstract class Protocol<SendRequestT extends Request, SendNotificationT e
             }
 
-            // Poll for task completion
-            while (true) {
-                // Get current task status
-                const task = await this.getTask({ taskId }, options);
-                yield { type: 'taskStatus', task };
-
-                // Check if task is terminal
-                if (isTerminal(task.status)) {
-                    if (task.status === 'completed') {
-                        // Get the final result
-                        const result = await this.getTaskResult({ taskId }, resultSchema, options);
-                        yield { type: 'result', result };
-                    } else if (task.status === 'failed') {
-                        yield {
-                            type: 'error',
-                            error: new McpError(ErrorCode.InternalError, `Task ${taskId} failed`)
-                        };
-                    } else if (task.status === 'cancelled') {
-                        yield {
-                            type: 'error',
-                            error: new McpError(ErrorCode.InternalError, `Task ${taskId} was cancelled`)
-                        };
-                    }
-                    return;
+            // Get initial task status
+            const initialTask = await this.getTask({ taskId }, options);
+            yield { type: 'taskStatus', task: initialTask };
+
+            // If already terminal, handle immediately
+            if (isTerminal(initialTask.status)) {
+                if (initialTask.status === 'completed') {
+                    const result = await this.getTaskResult({ taskId }, resultSchema, options);
+                    yield { type: 'result', result };
+                } else if (initialTask.status === 'failed') {
+                    yield {
+                        type: 'error',
+                        error: new McpError(ErrorCode.InternalError, `Task ${taskId} failed`)
+                    };
+                } else if (initialTask.status === 'cancelled') {
+                    yield {
+                        type: 'error',
+                        error: new McpError(ErrorCode.InternalError, `Task ${taskId} was cancelled`)
+                    };
                 }
-
-                // Wait before polling again
-                const pollInterval = task.pollInterval ?? this._options?.defaultTaskPollInterval ?? 1000;
-                await new Promise(resolve => setTimeout(resolve, pollInterval));
-
-                // Check if cancelled
-                options?.signal?.throwIfAborted();
+                return;
             }
+
+            // Call tasks/result - this will:
+            // 1. Deliver any queued messages (elicitation, sampling) via SSE
+            // 2. Block until task becomes terminal
+            // 3. Return final result
+            const result = await this.getTaskResult({ taskId }, resultSchema, options);
```

Comment:
The problem is why the server never returns a `input_required` state?
It should first return a `working` and then an `input_required`.

I think the problem is that the server-side queue should support prepend messages, where when there is an `input_required` message that needs to be processed, it should change the task's status to `input_required` and then the client's `tasks/get` will read that `input_required` and then pull the requests with `tasks/result`.

So this change is not valid AFAICS.

Reply:
Without making a call to `tasks/result` here, the client ends up in an endless loop (see example in the description table) because there's no way to actually complete the elicitation.

The server _does_ return a `input_required` result - we just don't end up doing anything other than continue polling because there's no handling here to do something else if `input_required` is reached.

However, there might be a case here to keep the loop and just add a case for input_required specifically, I'll try that.

Reply:
The client should switch to `tasks/result` after it sees the task status change to `input_required`, instead of calling `tasks/result` in the first place, but yes we can anyway:)